src/puppetlabs/puppetdb/http/fact_names.clj

```diff
@@ -1,40 +1,44 @@
 (ns puppetlabs.puppetdb.http.fact-names
-  (:require [puppetlabs.puppetdb.query.facts :as f]
+  (:require [puppetlabs.puppetdb.query.paging :as paging]
+            [puppetlabs.puppetdb.http.query :as http-q]
+            [net.cgrand.moustache :refer [app]]
             [puppetlabs.puppetdb.query-eng :refer [produce-streaming-body]]
-            [puppetlabs.puppetdb.query.paging :as paging]
-            [puppetlabs.puppetdb.jdbc :refer [with-transacted-connection]]
+            [clojure.walk :refer [keywordize-keys]]
             [puppetlabs.puppetdb.utils :refer [assoc-when]]
-            [net.cgrand.moustache :refer [app]]
             [puppetlabs.puppetdb.middleware :refer [verify-accepts-json
                                                     validate-query-params
-                                                    wrap-with-paging-options]]
-            [puppetlabs.puppetdb.http :refer [query-result-response]]))
+                                                    wrap-with-paging-options]]))
 
-(defn query-app
-  ([version]
-   (app
-     [&]
-     {:get (fn [{:keys [params globals paging-options]}]
-             (let [paging-options (assoc-when paging-options
-                                              :order_by [[:name :ascending]])]
+(defn query-route
+  "Returns a route for querying PuppetDB that supports the standard
+   paging parameters. Also accepts GETs and POSTs. Composes
+   `optional-handlers` with the middleware function that executes the
+   query."
+  [entity version param-spec & optional-handlers]
+  (app
+    (http-q/extract-query param-spec)
+    (apply comp
+           (fn [{:keys [params globals puppetdb-query]}]
+             (let [puppetdb-query (if (nil? (:order_by puppetdb-query))
+                                    (assoc puppetdb-query :order_by [[:name :ascending]])
+                                    puppetdb-query)]
                (produce-streaming-body
-                 :fact-names
+                 entity
                  version
-                 (params "query")
-                 paging-options
+                 (http-q/validate-distinct-options! (merge (keywordize-keys params) puppetdb-query))
                  (:scf-read-db globals)
-                 (:url-prefix globals))))})))
+                 (:url-prefix globals))))
+           optional-handlers)))
 
 (defn routes
-  [query-app]
-  (app
-    []
-    (verify-accepts-json query-app)))
+  [version]
+  (let [param-spec {:optional paging/query-params}]
+    (app
+      []
```

Comment:
no, I think we could definitely do that but I judged it to be unrelated to POST support. Worth a ticket IMO.